### PR TITLE
Use New Fideslang DSR-updated models [#1761]

### LINKED ISSUE #1761: Update Fides to use the Fideslang models
❗ DEPENDS ON ethyca/fideslang#94 ❗ 

Now that all of the DSR related metadata has been upstreamed into fideslang the codebase needs to be updated reference all the models from fideslang instead.

There will likely need to be some updates to existing dataset yaml files to no longer use `fidesops_meta`

Acceptance criteria
- Remove all of the code from [dataset.py](https://github.com/ethyca/fides/blob/ff4a4ca7b674e1dbca7c611094dffdb1c47c0f16/src/fides/api/ops/schemas/dataset.py) that was upstreamed into fideslang
- Update the codebase to use all the models from fideslang.

### PR DESCRIPTION
**Pull request by pattisdr**
❗ Blocked by https://github.com/ethyca/fideslang/pull/95
❗ Contains data migration

Closes #1761 

#### Installed Requirements
- _Temporarily_ point fideslang to the unmerged branch here  https://github.com/ethyca/fideslang/pull/95 while that PR remains unmerged. Revert before merge.
- _Temporarily_ install git in the Dockerfiles prior to installing requirements. Revert before merge.

#### Dataset yaml files
- Updated `redis_dataset.yml` (which is a dataset describing our redis cache) to not have data_categories at the object level when there are data_categories at the nested field level.  This is a constraint introduced from the fidesops-side to keep there from being data category conflicts in nested data structures when we're filtering data on data category before building the DSR package
- Updated the dataset yamls (for Database datasets, not saas datasets to have `fides_meta`.) Fideslang supports both in the request, but any fidesops_meta supplied is converted to fides_meta.  We have both in our yaml files right now.
- Update the `db_dataset.yml` which maps our fides database to have a `fides_meta` instead of a `fidesctl_meta` field on the ctl_datasets table

#### Replacing Fidesops Pydantic schemas with new Fideslang ones
- Use fideslang `FidesKey` in favor of `FidesOpsKey` - FidesKey now allows '>' '<' characters (useful for saas templates) and also bad validation throws a ValueError which is more easily picked up by FastAPI and turned into a ValidationError.
- Remove FidesopsDataset in favor of the new Fideslang Dataset which has the important attributes from FidesopsDataset
- Rename `Dataset` to be `GraphDataset` to not get confused with the Fideslang Dataset.  
- At the *code level* - adjust all fidesops_meta variables (at the Dataset, DatasetCollection, and DatasetField levels) to use `fides_meta` instead.  However, leave `fidesops_meta` keys in the saas-related yaml files for now.  If fidesops_meta is supplied, Fideslang converts it to fides_meta for backwards compatibility.
- Remove FidesopsDatasetReference in favor of identical Fideslang FidesDatasetReference

#### Validation
- Add new data category validation for the PATCH /datasetconfig endpoints where we validate against data_categories in the database instead of just the static default taxonomy as users could have added other data categories.  When running a DSR, the `dataset_config.dataset` (which is a Fideslang Dataset) is converted into a GraphDataset, and all the GraphDatasets are combined to make a `DatasetGraph`
- Because FidesKey validation errors now throw a FidesValidationError that inherits from ValueError instead of exception, adjust a couple of locations -ctl side where model validation is now throwing a ValidationError.

#### Database migration
- Update the `ctl_datasets.fidesctl_meta` to be named `fides_meta`

### Steps to Confirm

* [ ] Same steps in https://github.com/ethyca/fideslang/pull/95.  General QA around CTL Datasets and Ops Connections is handy.  Loading ops datasets into the db with the old fidesops_meta key and verifying that it is retrieved and serialized with fides_meta is a good check.  Main thing, both Ops DatasetConfig.datasets and Ctl ctl_datasets are being serialized with the same model now.

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation Updated:
  * [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  * [ ] documentation issue created (tag docs-team to complete issue separately)
* [ ] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created: https://github.com/ethyca/fides/issues/2016, https://github.com/ethyca/fideslang/issues/96, https://github.com/ethyca/fides/issues/2017
* [x] Update `CHANGELOG.md`

### Description Of Changes

Update Fides to use updated Fideslang modes (mostly changes on the -ops side) now that Fideslang has been updated so it's base models can support ops code.

We used to have two repositories: `fidesctl` and `fidesops` that were both using Fideslang `Dataset` concepts.  Fidesops added more items to subclassed Datasets to support DSRs.  We're now moving some of these fields to Fideslang, so we can share the same schemas between both -ctl and -ops usages.

### CODE REVIEW

File: tests/ops/api/v1/endpoints/test_dataset_endpoints.py
Comment (pattisdr) [inline tests/ops/api/v1/endpoints/test_dataset_endpoints.py:195]: Validation of field length was moved to Fideslang. It was updated to use a built-in Pydantic validator instead of our own custom validation to enforce positive non-zero models, so the error message has changed.

File: tests/ops/api/v1/endpoints/test_dataset_endpoints.py
Comment (pattisdr) [inline tests/ops/api/v1/endpoints/test_dataset_endpoints.py:291]: Updated validation was added to check data_categories against those in the database, not those in the static default data categories list.  To do this, I couldn't use a regular pydantic validator, but needed to do this after the fact, when I could access a database session.  So the response has adjusted as the request body itself isn't being validated here.

**Comment by pattisdr** [comment]
My last failing tests are in tests/ops/service/saas_request/test_saas_request_override_factory.py::TestSaasRequestOverrideFactory. 

It's odd because that file passes if you run it directly by itself, but if you run `nox -s "pytest_ops(unit)"` it fails.

File: src/fides/api/ops/schemas/dataset.py
Comment (pattisdr) [inline src/fides/api/ops/schemas/dataset.py:29]: Adding new validation on data category that looks in the database to see if data categories match ones defined there, instead of matching against the default data categories.

As a backup, if there's nothing in the db, which shouldn't be the case, matching against the static list.

File: src/fides/api/ops/schemas/dataset.py
Comment (pattisdr) [inline src/fides/api/ops/schemas/dataset.py:122]: Most of these removed concepts have moved to Fideslang.

File: src/fides/api/ops/schemas/shared_schemas.py
Comment (pattisdr) [inline src/fides/api/ops/schemas/shared_schemas.py:24]: Getting rid of FidesOpsKey in favor of an updated FidesKey.  

I've updated Fides Key to allow ">" "<" characters pending approval.  Also the error that FidesKey throws has been updated to be a subclass of a ValueError so can be converted into a pydantic ValidationError. 


The only behavior I didn't move to FidesKey was allowing the value argument to be Optional in FidesKey.value.  I think on our end we can just use Optional[FidesKey] for this.

File: src/fides/api/ops/service/saas_request/saas_request_override_factory.py
Comment (pattisdr) [inline src/fides/api/ops/service/saas_request/saas_request_override_factory.py:136]: This was causing some failing tests on this branch.  I think a '!=' comparison is recommended here anyway. We just want to see if the values are equal, not if they are stored in the same place in memory.

File: src/fides/api/ops/graph/config.py
Comment (pattisdr) [inline src/fides/api/ops/graph/config.py:494]: Renamed to not confuse with Fideslang Dataset, which we're using now in "ops" in favor of "FidesopsDataset"

To run a DSR, a Fideslang Dataset (stored in datasetconfig.dataset) is converted into a GraphDataset. Many GraphDatasets make up a DatasetGraph. The DatasetGraph is what we pass to `run_access_request` or `run_erasure_request`

**Comment by pattisdr** [comment]
@ethyca/docs-authors This work is changing `fidesops_meta` across the codebase to be `fides_meta` instead, although `fidesops_meta` is still supported for backwards compatibility.  I know this attribute is scattered throughout the docs -

File: src/fides/api/ops/service/connectors/saas/connector_registry_service.py
Comment (TheAndrewJackson) [inline src/fides/api/ops/service/connectors/saas/connector_registry_service.py:57]: Does it make sense to add this try/except block to the other decorators? My guess is that they also swallow the exceptions. That can definitely be a follow up ticket since it's more to test/QA

File: src/fides/api/ops/util/saas_util.py
Comment (TheAndrewJackson) [inline src/fides/api/ops/util/saas_util.py:121]: I think this may need to be updated

```suggestion
    Merges all Collections and Fields from the config_dataset into the GraphDataset.
```

File: tests/ops/task/traversal_data.py
Comment (TheAndrewJackson) [inline tests/ops/task/traversal_data.py:520]: small nit. I _think_ this is right. Feel free to ignore if it's wrong
```suggestion
def manual_graph_dataset(db_name: str, postgres_db_name) -> GraphDataset:
```

File: tests/ops/task/traversal_data.py
Comment (TheAndrewJackson) [inline tests/ops/task/traversal_data.py:521]: Similar thing here
```suggestion
    """Manual graph dataset depending on upstream postgres collection and pointing to a node in a downstream
```

File: data/dataset/bigquery_example_test_dataset.yml
Comment (TheAndrewJackson) [reply data/dataset/bigquery_example_test_dataset.yml:14]: Good idea. Should/can the saas templates be converted over to use `fides_meta` in a future PR down the line in the unified fides resources work?

File: src/fides/api/ops/schemas/dataset.py
Comment (TheAndrewJackson) [reply src/fides/api/ops/schemas/dataset.py:29]: Would it make sense to add an `INFO` log saying what `DataCategory` didn't have any categories in the database? If it's a situation that shouldn't happen it might be worth making a note of it.

**Comment by TheAndrewJackson** [comment]
Does it make sense to merge this into a feature branch instead of directly into main? This epic won't be finished until a few more PRs are merge in. Another fides release may happen before that point and release some of the Unified Resources epic early.

File: data/dataset/bigquery_example_test_dataset.yml
Comment (pattisdr) [reply data/dataset/bigquery_example_test_dataset.yml:14]: From this comment, it sounded like we wanted to preserve the backwards compatibility of this for now https://github.com/ethyca/fideslang/issues/94#issuecomment-1341175140

**Comment by pattisdr** [comment]
Hi @TheAndrewJackson thanks for the feature branch suggestion. I'd tried to make this PR standalone but we've discussed some other breaking changes in subsequent PR's.  So I think a feature branch makes sense and we can really check to make sure it's been solidly QA'd on the FE and BE before merge.

File: tests/ops/task/traversal_data.py
Comment (pattisdr) [reply tests/ops/task/traversal_data.py:520]: sure I can change the name of the test function!

File: src/fides/api/ops/util/saas_util.py
Comment (pattisdr) [reply src/fides/api/ops/util/saas_util.py:121]: This code comment to me looks like it's referring to the variable names not the data types of those variable names. I can wrap them in quotes in the docstring -

File: src/fides/api/main.py
Comment (pattisdr) [reply src/fides/api/main.py:233]: I believe this was your update! I left it in, seemed helpful.

File: src/fides/api/ops/service/connectors/saas/connector_registry_service.py
Comment (pattisdr) [reply src/fides/api/ops/service/connectors/saas/connector_registry_service.py:57]: ah I actually meant to revert this change you'd added - I wasn't seeing the swallowed exceptions on my end:

If I create an error with my datasets:

```bash
datastores specified: postgres
configuring infrastructure for postgres
[+] Running 4/5
 ⠿ Network fides_default               Created                                                                                                                            0.0s
 ⠿ Container fides-redis-1             Healthy                                                                                                                           17.4s
 ⠿ Container fides-fides-db-1          Healthy                                                                                                                           17.4s
 ⠿ Container fides-postgres_example-1  Healthy                                                                                                                           17.4s
 ⠿ Container fides-fides-1             Waiting
```

If I go into the  fides docker container with the try/except removed:
```2022-12-13 19:50:32.640 [INFO] (main:setup_server:228): Validating SaaS connector templates...
2022-12-13 19:50:33.860 [ERROR] (main:setup_server:235): Error occurred during SaaS connector template validation: 1602 validation errors for ConnectorRegistry
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 0 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 1 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 3 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 5 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 7 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 10 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 12 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 14 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 16 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 17 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 19 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 20 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 21 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 22 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 23 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 26 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 27 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 28 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 29 -> fides_meta -> blah
  field required (type=value_error.missing)
__root__ -> adobe_campaign -> dataset -> collections -> 0 -> fields -> 30 -> fides_meta -> blah
  field required (type=value_error.missing)
...
```

Comment (TheAndrewJackson) [reply src/fides/api/ops/service/connectors/saas/connector_registry_service.py]: I'm pretty sure I did but it's possible I wasn't looking in the right place. If it's no longer an issue then I'm okay with it being removed 👍

**Review by TheAndrewJackson — APPROVED** [review]

File: requirements.txt
Comment (pattisdr) [reply requirements.txt]: Leaving for now in the feature branch as we haven't released fideslang yet but will make sure to bump this before it's merged to main.

**Comment by pattisdr** [comment]
Thanks for your review @TheAndrewJackson! Merging into the feature branch  -

**Pull request merged**
